File: src/routes/extensions.js
/**
 * @swagger
 * tags:
 *   name: Extensions
 *   description: FusionPBX extensions and registrations
 */

const express = require('express');
const { query, param } = require('express-validator');
const { authenticate } = require('../middleware/auth');
const { validate } = require('../middleware/validate');
const dbService = require('../services/dbService');
const fusionpbxService = require('../services/fusionpbxService');

const router = express.Router();
router.use(authenticate);

/**
 * @swagger
 * /api/extensions:
 *   get:
 *     summary: Get all extensions for a domain
 *     tags: [Extensions]
 *     security:
 *       - ApiKeyAuth: []
 *       - BearerAuth: []
 *     parameters:
 *       - in: query
 *         name: domain
 *         required: true
 *         schema:
 *           type: string
 *     responses:
 *       200:
 *         description: List of extensions
 */
router.get(
  '/',
  [query('domain').notEmpty().withMessage('domain is required')],
  validate,
  async (req, res, next) => {
    try {
      const extensions = await dbService.getExtensions(req.query.domain);
      res.json({ extensions, count: extensions.length });
    } catch (err) {
      next(err);
    }
  }
);

/**
 * @swagger
 * /api/extensions/{extension}:
 *   get:
 *     summary: Get a specific extension by number
 *     tags: [Extensions]
 *     security:
 *       - ApiKeyAuth: []
 *       - BearerAuth: []
 *     parameters:
 *       - in: path
 *         name: extension
 *         required: true
 *         schema:
 *           type: string
 *       - in: query
 *         name: domain
 *         required: true
 *         schema:
 *           type: string
 *     responses:
 *       200:
 *         description: Extension details
 *       404:
 *         description: Extension not found
 */
router.get(
  '/:extension',
  [
    param('extension').notEmpty(),
    query('domain').notEmpty().withMessage('domain is required'),
  ],
  validate,
  async (req, res, next) => {
    try {
      const ext = await dbService.getExtensionByNumber(req.params.extension, req.query.domain);
      if (!ext) return res.status(404).json({ error: 'Extension not found' });
      res.json({ extension: ext });
    } catch (err) {
      next(err);
    }
  }
);

/**
 * @swagger
 * /api/extensions/registrations:
 *   get:
 *     summary: Get registered SIP endpoints (online/offline status)
 *     tags: [Extensions]
 *     security:
 *       - ApiKeyAuth: []
 *       - BearerAuth: []
 *     parameters:
 *       - in: query
 *         name: domain
 *         schema:
 *           type: string
 *     responses:
 *       200:
 *         description: Registration list
 */
router.get('/registrations', async (req, res, next) => {
  try {
    const regs = await fusionpbxService.getRegistrations(req.query.domain);
    res.json({ registrations: regs });
  } catch (err) {
    next(err);
  }
});

module.exports = router;

File: python/routers/extensions.py
from fastapi import APIRouter, Depends, HTTPException, Query
from typing import Optional
from deps.auth import require_auth
from services.db_service import db_service
from services.fusionpbx_service import fusionpbx_service

router = APIRouter(prefix='/api/extensions', tags=['Extensions'])


def _effective_domain(user: dict, requested: Optional[str] = None) -> Optional[str]:
    if user.get('source') == 'user':
        return user.get('domain')
    return requested


@router.get('/registrations')
async def get_registrations(
    domain: Optional[str] = Query(None),
    _user=Depends(require_auth),
):
    regs = await fusionpbx_service.get_registrations(_effective_domain(_user, domain))
    return {'registrations': regs}


@router.get('')
@router.get('/')
async def get_extensions(
    domain: str = Query(...),
    _user=Depends(require_auth),
):
    extensions = await db_service.get_extensions(_effective_domain(_user, domain))
    return {'extensions': extensions, 'count': len(extensions)}


@router.get('/{extension}')
async def get_extension(
    extension: str,
    domain: str = Query(...),
    _user=Depends(require_auth),
):
    ext = await db_service.get_extension_by_number(extension, _effective_domain(_user, domain))
    if not ext:
        raise HTTPException(404, 'Extension not found')
    return {'extension': ext}
